File: leet-code/406. Queue Reconstruction by Height/QueueReconstructionByHeight406_2.java
package com.example;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.PriorityQueue;

public class QueueReconstructionByHeight406_2 {
    public int[][] reconstructQueue(int[][] people) {
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[0] == b[0] ? a[1] - b[1] : b[0] - a[0]);
        pq.addAll(Arrays.asList(people));
        List<int[]> answer = new ArrayList<>();
        while (!pq.isEmpty()) {
            int[] item = pq.poll();
            answer.add(item[1], item);
        }
        return answer.toArray(new int[0][0]);
    }
}

File: leet-code/406. Queue Reconstruction by Height/QueueReconstructionByHeight406_2.kt
package com.example

import java.util.PriorityQueue

class QueueReconstructionByHeight406_2 {
    fun reconstructQueue(people: Array<IntArray>): Array<IntArray> {
        val pq = PriorityQueue<IntArray> { p1, p2 ->
            when {
                p1[0] == p2[0] -> p1[1] - p2[1]
                else -> p2[0] - p1[0]
            }
        }
        pq.addAll(people)

        val answer = mutableListOf<IntArray>()
        while (!pq.isEmpty()) {
            val item = pq.poll()
            answer.add(item[1], item)
        }
        return answer.toTypedArray()
    }
}
